pkg/policy/rule.go

```diff
@@ -162,6 +153,20 @@ func mergePortProto(ctx *SearchContext, existingFilter, filterToMerge *L4Filter,
 			// Merge isRedirect flag
 			l7Rules.isRedirect = l7Rules.isRedirect || newL7Rules.isRedirect
 
+			// Merge listener reference
+			if l7Rules.Listener == "" || newL7Rules.Listener == "" {
+				if newL7Rules.Listener != "" {
+					l7Rules.Listener = newL7Rules.Listener
+				}
+			} else if newL7Rules.Listener != l7Rules.Listener &&
+				newL7Rules.Priority == l7Rules.Priority {
+				ctx.PolicyTrace("   Merge conflict: mismatching CiliumEnvoyConfig listeners %v/%v with the same priority %d\n", newL7Rules.Listener, l7Rules.Listener, l7Rules.Priority)
+				return fmt.Errorf("cannot merge conflicting CiliumEnvoyConfig Listeners (%v/%v) with the same priority (%d)", newL7Rules.Listener, l7Rules.Listener, l7Rules.Priority)
+			}
+			if newL7Rules.Listener != l7Rules.Listener && newL7Rules.Priority < l7Rules.Priority {
+				l7Rules.Listener = newL7Rules.Listener
+			}
+
```

Comment:
What about using a helper like the following:

```
func mergeListenerReference(ctx *SearchContext, l7Rules, newL7Rules *PerSelectorPolicy) error {
	if newL7Rules.Listener == "" {
		return nil
	}

	if l7Rules.Listener == "" {
		l7Rules.Listener = newL7Rules.Listener
		return nil
	}

	if l7Rules.Listener == newL7Rules.Listener || newL7Rules.Priority > l7Rules.Priority {
		return nil
	}

	if newL7Rules.Priority < l7Rules.Priority {
		l7Rules.Listener = newL7Rules.Listener
		return nil
	}

	// conflict: newL7Rules.Priority == l7Rules.Priority
	ctx.PolicyTrace("   Merge conflict: mismatching CiliumEnvoyConfig listeners %v/%v with the same priority %d\n", newL7Rules.Listener, l7Rules.Listener, l7Rules.Priority)
	return fmt.Errorf("cannot merge conflicting CiliumEnvoyConfig Listeners (%v/%v) with the same priority (%d)", newL7Rules.Listener, l7Rules.Listener, l7Rules.Priority)
}
```

Also, I think it'd be better to add a test for this.
Finally, why the priority is not overwritten when the newL7Rules Listener is written over the existing one?